sgo/types/lookup.go

```diff
@@ -294,6 +293,36 @@ func assertableTo(V *Interface, T Type) (method *Func, wrongType bool, mustOptio
 	return
 }
 
+// typeMustOptional reports whether a type T recursively needs to be wrapped in an optional.
+// If T is Array or Slice, the underlying type will be checked. If T is Optional, it will not
+// need to be wrapped in an optional unless the underlying type is a Map. In that case, further
+// checks are needed for the key and value type of the map.
+func typeMustOptional(T Type) bool {
```

Comment:
Yes, why not. Plus, for better error reporting, you could return the types that needed to be wrapped in optionals but aren't instead of just true/false.